lib/Dialect/FIRRTL/Transforms/MaterializeDebugInfo.cpp

```diff
@@ -0,0 +1,111 @@
+//===- MaterializeDebugInfo.cpp - DI materialization ----------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "circt/Dialect/Debug/DebugOps.h"
+#include "circt/Dialect/FIRRTL/FIRRTLOps.h"
+#include "circt/Dialect/FIRRTL/FIRRTLTypes.h"
+#include "mlir/IR/Attributes.h"
+#include "mlir/IR/Builders.h"
+#include "llvm/ADT/StringExtras.h"
+#include "llvm/Support/Parallel.h"
+
+using namespace mlir;
+using namespace circt;
+using namespace firrtl;
+
+namespace {
+struct MaterializeDebugInfoPass
+    : public MaterializeDebugInfoBase<MaterializeDebugInfoPass> {
+  void runOnOperation() override;
+  void materializeVariable(OpBuilder &builder, StringAttr name, Value value);
+  Value convertToDebugAggregates(OpBuilder &builder, Value value);
+};
+} // namespace
+
+void MaterializeDebugInfoPass::runOnOperation() {
+  auto module = getOperation();
+  auto builder = OpBuilder::atBlockBegin(module.getBodyBlock());
+
+  // Create DI variables for each port.
+  for (const auto &[port, value] :
+       llvm::zip(module.getPorts(), module.getArguments())) {
+    materializeVariable(builder, port.name, value);
+  }
+
+  // Create DI variables for each declaration in the module body.
+  module.walk([&](Operation *op) {
+    TypeSwitch<Operation *>(op).Case<WireOp, NodeOp, RegOp, RegResetOp>(
+        [&](auto op) {
+          builder.setInsertionPointAfter(op);
+          materializeVariable(builder, op.getNameAttr(), op.getResult());
+        });
+  });
+}
+
+/// Materialize debug variable ops for a value.
+void MaterializeDebugInfoPass::materializeVariable(OpBuilder &builder,
+                                                   StringAttr name,
+                                                   Value value) {
+  if (!name || isUselessName(name.getValue()))
+    return;
+  if (name.getValue().starts_with("_"))
+    return;
+  if (auto dbgValue = convertToDebugAggregates(builder, value))
+    builder.create<debug::VariableOp>(value.getLoc(), name, dbgValue);
+}
+
+/// Unpack all aggregates in a FIRRTL value and repack them as debug aggregates.
+/// For example, converts a FIRRTL vector `v` into `dbg.array [v[0],v[1],...]`.
+Value MaterializeDebugInfoPass::convertToDebugAggregates(OpBuilder &builder,
+                                                         Value value) {
+  return TypeSwitch<Type, Value>(value.getType())
```

Comment:
nit: Please use FIRRTLTypeSwitch to handle type aliases.
```suggestion
  return FIRRTLTypeSwitch<Type, Value>(value.getType())
```